File: src/utils/server_identifier.py
import aiohttp
from typing import Dict, Any

SERVER_SIGNATURES = {
    'cloudflare': ['cloudflare', 'cf-ray'],
    'cloudfront': ['cloudfront', 'amazon'],
    'google': ['gws', 'Google Frontend', 'Google Cloud', 'GFE'],
    'fastly': ['fastly'],
    'akamai': ['akamai', 'akamaighost'],
    'nginx': ['nginx'],
    'apache': ['apache'],
    'microsoft': ['microsoft-iis', 'azure'],
    'amazon': ['amazon', 'aws', 'ec2'],
    'cdn': ['cdn', 'cdnjs', 'jsdelivr']
}

async def identify_server(host: str, session: aiohttp.ClientSession, timeout: float) -> Dict[str, Any]:
    try:
        async with session.head(f'https://{host}') as response:
            headers = dict(response.headers)
            server_info = {
                'type': 'unknown',
                'headers': {}
            }

            # Store relevant headers
            for header in ['server', 'x-powered-by', 'x-server', 'via']:
                if header in headers:
                    server_info['headers'][header] = headers[header]

            # Check all headers for server signatures
            headers_str = ' '.join(str(v).lower() for v in headers.values())
            
            for provider, signatures in SERVER_SIGNATURES.items():
                if any(sig.lower() in headers_str for sig in signatures):
                    server_info['type'] = provider
                    break

            return server_info
    except Exception as e:
        return {
            'type': 'unknown',
            'error': str(e)
        }

File: src/serverIdentifier.js
import fetch from 'node-fetch';

const SERVER_SIGNATURES = {
  'cloudflare': ['cloudflare', 'cf-ray'],
  'cloudfront': ['cloudfront'],
  'google': ['gws', 'Google Frontend'],
  'fastly': ['fastly'],
  'akamai': ['akamai'],
  'nginx': ['nginx'],
  'apache': ['apache']
};

export async function identifyServer(host) {
  try {
    const response = await fetch(`https://${host}`, {
      method: 'HEAD'
    });

    const headers = response.headers;
    const server = headers.get('server');
    const serverInfo = {
      type: 'unknown',
      headers: {}
    };

    // Check server header
    if (server) {
      serverInfo.headers.server = server;
    }

    // Check for specific headers that indicate the server type
    for (const [provider, signatures] of Object.entries(SERVER_SIGNATURES)) {
      for (const signature of signatures) {
        if (
          server?.toLowerCase().includes(signature.toLowerCase()) ||
          Array.from(headers.entries()).some(([key, value]) => 
            value.toLowerCase().includes(signature.toLowerCase())
          )
        ) {
          serverInfo.type = provider;
          break;
        }
      }
    }

    return serverInfo;
  } catch (error) {
    return {
      type: 'unknown',
      error: error.message
    };
  }
}